src/Network/Transport/TCP.hs

```diff
@@ -855,16 +854,33 @@ apiCloseEndPoint transport evs ourEndPoint =
 -- the transport down. We must be careful to close the socket when a (possibly
 -- asynchronous, ThreadKilled) exception occurs. (If an exception escapes from
 -- handleConnectionRequest the transport will be shut down.)
-handleConnectionRequest :: TCPTransport -> N.Socket -> IO ()
-handleConnectionRequest transport sock = handle handleException $ do
+handleConnectionRequest :: TCPTransport -> (N.Socket, N.SockAddr) -> IO ()
+handleConnectionRequest transport (sock, sockAddr) = handle handleException $ do
     when (tcpNoDelay $ transportParams transport) $
       N.setSocketOption sock N.NoDelay 1
     when (tcpKeepAlive $ transportParams transport) $
       N.setSocketOption sock N.KeepAlive 1
     forM_ (tcpUserTimeout $ transportParams transport) $
       N.setSocketOption sock N.UserTimeout
+    -- Get the OS-determined host and port.
+    (actualHost, actualPort) <-
+      decodeSockAddr sockAddr >>=
+        maybe (throwIO (userError "handleConnectionRequest: invalid socket address")) return
+    -- Receive our EndPointId from the peer.
     ourEndPointId <- recvWord32 sock
+    -- Let the peer tell us their EndPointAddress.
     theirAddress  <- EndPointAddress . BS.concat <$> recvWithLength sock
+    (theirHost, theirPort, theirEndPointId)
+      <- maybe (throwIO (userError "handleConnectionRequest: peer gave malformed address"))
+               return
+               (decodeEndPointAddress theirAddress)
+    -- If the OS-determined host doesn't match the host that the peer gave us,
+    -- then we have no choice but to reject the connection. It's because we
+    -- use the EndPointAddress to key the remote end points (localConnections)
+    -- and we don't want to allow a peer to deny service to other peers by
+    -- claiming to have their host and port.
+    unless (theirHost == actualHost) $ do
+      throwIO (userError "handleConnectionRequest: reported host does not match actual host")
```

Comment:
Please, include the values of `theirHost` and `actualHost` in the error message.